### PR TITLE
IGNITE-14169 .NET: Add NamespacePrefix and NamespaceToLower to BinaryBasicNameMapper

### PR DESCRIPTION
**Pull request by nizhikov**
Add `NamespacePrefix` and `NamespaceToLower` properties to `BinaryBasicNameMapper` to allow for automatic conversion between different type naming styles.

Currently, `BinaryBasicNameMapper[isSimpleName=false]` assumes that type names equals both in Java and .NET.

  * Java type - `ru.company.Model`
  * .NET type - `ru.company.Model`

However, .NET naming conventions [1] differ from Java [2]
So, if a user follows both conventions then names will be

  * Java type - `ru.company.Model`
  * .NET type - `Company.Model`

Users can implement this behavior in a custom name mapper, but custom name mapper will disable `RegisterSameJavaType` mode.

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @ptupitsyn 
Can you, please, take a look at my changes?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/TypeNameParser.cs [hunk 1]
```diff
@@ -30,19 +31,23 @@ internal class TypeNameParser
...
         /// Initializes a new instance of the <see cref="TypeNameParser" /> class.
         /// </summary>
-        private TypeNameParser(string typeName, ref int pos)
+        private TypeNameParser(string typeName, ref int pos, bool forceJavaNamingConventions)
         {
-            _typeName = typeName;
+            _typeNameArr = typeName.ToCharArray();
             _start = pos;
             _pos = _start;
+            _forceJavaNamingConventions = forceJavaNamingConventions;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/TypeNameParser.cs:50]: Separation of concerns: `TypeNameParser` should not care about `forceJavaNamingConventions`, it just parses the .NET type name. We should perform the `ToLower` call elsewhere.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/TypeNameParser.cs (hunk 1 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/TypeNameParser.cs:50]: FIxed.

**Comment by nizhikov** [comment]
Tests - https://ci.ignite.apache.org/viewLog.html?buildId=5874215&buildTypeId=IgniteTests24Java8_RunAllNet

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 2]
```diff
@@ -48,7 +48,7 @@ public class BinaryBasicNameMapper : IBinaryNameMapper
         /// <summary>
         /// Gets the type name.
         /// </summary>
-        public string GetTypeName(string name)
+        public virtual string GetTypeName(string name)
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:51]: @nizhikov please see the [dev list](http://apache-ignite-developers.2346864.n4.nabble.com/NET-naming-convention-RegisterSameJavaType-mode-td51386.html) - making those methods `virtual` does not seem to be a good idea.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 2 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs]: I reworked the solution according to your proposals on the dev-list.

`NamespaceToLower` and `JavaDomain` flags added to `BinaryBasicNameMapper`.
Please, take a look.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 3]
```diff
@@ -45,6 +46,20 @@ public class BinaryBasicNameMapper : IBinaryNameMapper
         /// </summary>
         public bool IsSimpleName { get; set; }
 
+        /// <summary>
+        /// Domain to be added as a first part of Java type name.
+        /// Java and .Net assumes usage of different naming conventions.
+        /// Java package name first part is a domain: com, org, ru, etc. and second part is a company name: acme, company, etc.
+        /// .Net naming conventions don't use domain and start with a company name.
+        /// </summary>
+        public string JavaDomain { get; set; }
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:55]: I would prefer the `NamespacePrefix` name for this property - seems to be more technically correct, it is not necessarily Java-specific.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 4]
```diff
@@ -45,6 +46,20 @@ public class BinaryBasicNameMapper : IBinaryNameMapper
         /// </summary>
         public bool IsSimpleName { get; set; }
 
+        /// <summary>
+        /// Domain to be added as a first part of Java type name.
+        /// Java and .Net assumes usage of different naming conventions.
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:51]: Here and below - it is spelled `.NET`, all upper case.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 5]
```diff
@@ -58,7 +73,12 @@ public string GetTypeName(string name)
             {
                 // Generics are rare, use simpler logic for the common case.
                 var res = IsSimpleName ? parsedName.GetName() : parsedName.GetNameWithNamespace();
-                
+
+                if (!IsSimpleName && NamespaceToLower)
+                {
+                    res = DoForceJavaNamingConventions(res, JavaDomain == null ? null : JavaDomain + '.');
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:79]: I don't think we need to add `.` here, let the user deal with this. Along with `NamespacePrefix` property name, this becomes a more general feature - you can have any prefix, with or without `.` at the end.

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:79]: `JavaDomain` is only added when `NamespaceToLower` is true - is that intentional?

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 6]
```diff
@@ -125,5 +157,27 @@ public string GetFieldName(string name)
...
+            if (javaDomain != null)
+            {
+                name = javaDomain + name;
+            }
+
+            var arr = name.ToCharArray();
+            var nameStart = name.LastIndexOf('.');
+
+            for (int j = 0; j < nameStart; j++)
+                arr[j] = Char.ToLower(arr[j], CultureInfo.CurrentCulture);
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:178]: `CultureInfo.InvariantCulture` seems to be a better fit here - we don't want to depend on local culture settings for namespace conversion.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 7]
```diff
@@ -125,5 +157,27 @@ public string GetFieldName(string name)
...
+        /// <param name="javaDomain">Java domain to be added.</param>
+        /// <returns></returns>
+        private static string DoForceJavaNamingConventions(string name, string javaDomain)
+        {
+            if (javaDomain != null)
+            {
+                name = javaDomain + name;
+            }
+
+            var arr = name.ToCharArray();
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:174]: I think we can simplify lowering logic if we add `TypeNameParser.GetNamespace` method, something like:
`p.GetNamespace.ToLower(CultureInfo.InvariantCulture) + '.' + p.GetName()`.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 4 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:51]: Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 3 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:55]: OK. Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 5 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:79]: > I don't think we need to add . here

OK. Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 6 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:178]: Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs [hunk 8]
```diff
@@ -87,7 +119,7 @@ public string GetFieldName(string name)
         /// <summary>
         /// Builds the type name.
         /// </summary>
-        private static StringBuilder BuildTypeName(TypeNameParser typeName, StringBuilder sb, 
+        internal static StringBuilder BuildTypeName(TypeNameParser typeName, StringBuilder sb,
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:122]:
```suggestion
        private static StringBuilder BuildTypeName(TypeNameParser typeName, StringBuilder sb, 
```

Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs:122]: Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs (hunk 7 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs]: Done.

**Comment by nizhikov** [comment]
@ptupitsyn Thanks for the review!

I fixed all your comments. Please, take a look one more time.

**Comment by nizhikov** [comment]
https://ci.ignite.apache.org/viewLog.html?buildId=5876312&tab=queuedBuildOverviewTab - tests.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryNameMapperTest.cs [hunk 9]
```diff
@@ -95,6 +96,71 @@ public void TestSimpleName()
...
+            Assert.AreEqual("URL", mapper.GetTypeName("Org.Company.URL"));
+        }
+
+        /// <summary>
+        /// Tests BinaryBasicNameMapperForJava and JavaDomain = "org".
+        /// </summary>
+        [Test]
+        public void TestFullNameForceJavaNamingConventionsWithDomain()
+        {
+            var mapper = new BinaryBasicNameMapper {NamespacePrefix = "org.", NamespaceToLower = true};
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryNameMapperTest.cs:149]: One more combination is missing, when `NamespacePrefix` is present, but `NamespaceToLower` is `false` - let's add a test method for this too.

**Review by ptupitsyn — COMMENTED** [review]
@nizhikov I've tried to simplify the name conversion a little bit, please check if my changes look good to you.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryNameMapperTest.cs (hunk 9 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryNameMapperTest.cs:149]: Fixed

**Review by ptupitsyn — APPROVED** [review]
Looks good to me :+1:

**Comment by nizhikov** [comment]
@ptupitsyn Thanks for the review! Appreciate it.

I will rerun .Net tests and merge this PR.

**Pull request merged**